### PR TITLE
VSCode Extension 2.0

### PR DESCRIPTION
**Pull request by Philipp15b**
Based on PR #12 by @umutdural:

> This pull request adds the initial version of the new VSCode extension that communicates with a server version of Caesar which can continuously verify files through the extension without using the CLI. The Caesar server also serves as a language server for the HeyVL ( which allows the display of HeyVL diagnostics in the editor).
> 
> The extension initially supports the verification of files with a Caesar: Verify command or it can verify files automatically on save.
> User Interface:
> 
>     Gutter Icons: Based on the verification result (Verified,Failed,Unknown), appropriate icons are displayed next to each (co)proc in the gutter.
>     Status Bar: A status bar displays the status of the server.
>     Inline Ghost Text for computed WP/WLP/ERT: Before each instruction if the line above is empty, the expectation of the program (w.r.t. the calculus chosen) after that instruction will be displayed as a ghost text (similar to comments).
> 
> All of the above functionalities can be toggled from the settings in VSCode.
>

### CODE REVIEW

File: src/ast/diagnostic.rs [hunk 1]
```diff
@@ -198,6 +198,43 @@ impl Span {
...
+    pub fn to_lsp_range(self, files: &Files) -> Option<lsp_types::Range> {
+        let file = files.get(self.file).unwrap();
+        let char_span = file.char_span(self);
+
+        let mut start_line = 0;
+        let mut start_offset = 0;
+        let mut end_line = 0;
+        let mut end_offset = 0;
+
+        for (i, c) in file.source.chars().enumerate() {
```
Comment (Philipp15b) [inline src/ast/diagnostic.rs:211]: @umutdural Can you say why you need to iterate over the source code here again?

Comment (umutdural) [reply src/ast/diagnostic.rs:211]: char_span gives the span in terms of characters from the beginning of the file. However VSCode API requires spans (ranges) in line and offset format. I needed to find which start and end lines the char span corresponds to, thus I needed to count the number of newline characters until the beginning and end of the char span.

Comment (Philipp15b) [reply src/ast/diagnostic.rs:211]: Ah I see. This functionality is available in ariadne already. If you use a modified version of [this code](https://github.com/moves-rwth/caesar/blob/cd06d796a41477e8877cda4744460aba07720b2a/src/ast/diagnostic.rs#L127-L137), you should be able to obtain this data as well. Main function is [ariadne's `get_offset_line`](https://docs.rs/ariadne/latest/ariadne/struct.Source.html#method.get_offset_line).

**Comment by Philipp15b** [comment]
Thanks @umutdural!

**Pull request merged**